# Sourcing Open Data: E-commerce Cosmetic Products

# Table of contents:
1. Importing libraries and Dataset
2. Data Check
3. Data Cleaning
4. Renaming Columns
5. Data Understanding

## 1. Importing libraries and Dataset

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os


In [2]:
#import path and  E-commerce Cosmetic Products dataset
path=r'E:\E-commerce Cosmetic Products'

In [3]:
# Load the dataset with a specified encoding
df_cos = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Ecommerce_cosmetic_dataset.csv'), encoding='latin1', index_col=False)

## 2. Data Check

In [4]:
df_cos.head(10)

product_name   website country  \
0  Carlton London Incense Eau da parfum, Premium ...  Flipkart   India   
1  CHARLENE SPRAY MIST PERFUME 30 - INTIMATE (PAC...  Flipkart   India   
2  CHARLENE SPRAY MIST PERFUME 30 - INTIMATE (PAC...  Flipkart   India   
3  DENVER Black Code Perfume - 60 | Eau de Parfum...  Flipkart   India   
4  Denver Hamilton Perfume - 100 | Long Lasting P...  Flipkart   India   
5  Engage Fantasia Perfume for Women, Long Lastin...  Flipkart   India   
6  Engage Verona Perfume for Women, Long Lasting,...  Flipkart   India   
7  Engage One Soul Gender-free Perfume, Unisex, L...  Flipkart   India   
8  Engage Indigo Skies Perfume for Men, Long Last...  Flipkart   India   
9  ENVY Men Perfume - 70 | Eau da parfum - For Me...  Flipkart   India   

  category subcategory                                         title-href  \
0     body     perfume  https://www.amazon.in/Carlton-London-Limited-I...   
1     body     perfume  https://www.amazon.in/CHARLENE-SPRAY-MIST-PERF...   
2     body     perfume  https://www.amazon.in/CHARLENE-SPRAY-MIST-PERF...   
3     body     perfume  https://www.amazon.in/DENVER-Black-Code-Perfum...   
4     body     perfume  https://www.amazon.in/Denver-Perfume-Hamilton-...   
5     body     perfume  https://www.amazon.in/Engage-Fantasia-Perfume-...   
6     body     perfume  https://www.amazon.in/Engage-Verona-Perfume-La...   
7     body     perfume  https://www.amazon.in/Engage-Gender-free-Perfu...   
8     body     perfume  https://www.amazon.in/Engage-Indigo-Perfume-La...   
9     body     perfume  https://www.amazon.in/ENVY-Men-Perfume-parfum-...   

   price           brand ingredients     form type  \
0  599.0  Carlton London         NaN  aerosol  NaN   
1  149.0        Charlene         NaN  aerosol  NaN   
2  298.0        Charlene         NaN  aerosol  NaN   
3  245.0          Denver         NaN  aerosol  NaN   
4  422.0          Denver         NaN  aerosol  NaN   
5  422.0          Engage         NaN  aerosol  NaN   
6  422.0          Engage         NaN  aerosol  NaN   
7  422.0          Engage         NaN  aerosol  NaN   
8  422.0          Engage         NaN  aerosol  NaN   
9  236.0            Envy         NaN  aerosol  NaN   

                                               color size rating noofratings  
0  Top Note: Orange Blossom, Blackberry | Heart N...  100    3.9          19  
1                                   Unit count type:   30    4.4       4,031  
2                                   Unit count type:   30    4.4       4,072  
3                                 Long-Lasting Scent   60    4.2          61  
4                                 Long-Lasting Scent  100    4.3         342  
5  Engage Fantasia is a premium, long lasting per...  100    4.4      13,101  
6  Inspired by citrus grooves, Engage Verona is a...  100    4.3         515  
7  Engage One Soul is a gender free and long last...  100    4.5       9,665  
8  Engage Indigo Skies is a premium, long lasting...  100    4.3          54  
9                                 Long-Lasting Scent   70    4.2         842

In [5]:
df_cos.shape

(12615, 15)

In [6]:
df_cos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12615 entries, 0 to 12614
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_name  12615 non-null  object 
 1   website       12615 non-null  object 
 2   country       12615 non-null  object 
 3   category      12615 non-null  object 
 4   subcategory   12615 non-null  object 
 5   title-href    12615 non-null  object 
 6   price         12298 non-null  float64
 7   brand         12615 non-null  object 
 8   ingredients   6600 non-null   object 
 9   form          12615 non-null  object 
 10  type          9934 non-null   object 
 11  color         10626 non-null  object 
 12  size          9449 non-null   object 
 13  rating        10548 non-null  object 
 14  noofratings   12156 non-null  object 
dtypes: float64(1), object(14)
memory usage: 1.4+ MB


## 3.Data Cleaning

##### I made a selection of which columns I keep and which ones I can remove for my project. I will first make a dataframe of the columns I keep before I do consistency checks and data cleaning.

In [7]:
# Columns'Product_name', 'title_hef' ,'ingredients' are removed
columns_to_keep = ['website', 'country', 'category', 'subcategory', 'price', 'brand', 'form', 'type', 'color', 'size', 'rating', 'noofratings']
df_cosmetic = df_cos[columns_to_keep]

In [8]:
df_cosmetic.shape

(12615, 12)

In [9]:
# check for mixed types
for col in df_cosmetic.columns.tolist():
  weird = (df_cosmetic[[col]].applymap(type) != df_cosmetic[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cosmetic[weird]) > 0:
    print (col)

type
color
size
rating
noofratings


C:\Users\negar\AppData\Local\Temp\ipykernel_19044\2371407318.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cosmetic[[col]].applymap(type) != df_cosmetic[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\negar\AppData\Local\Temp\ipykernel_19044\2371407318.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cosmetic[[col]].applymap(type) != df_cosmetic[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\negar\AppData\Local\Temp\ipykernel_19044\2371407318.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cosmetic[[col]].applymap(type) != df_cosmetic[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\negar\AppData\Local\Temp\ipykernel_19044\2371407318.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_cosmetic[[col]].applymap(type) != df_cosmetic[[col]].iloc[0].apply(type)).any(axis =

In [10]:
df_cosmetic['type'].value_counts(dropna = False)

type
all                               5158
NaN                               2681
matte                              572
combination, dry, normal, oily     528
shimmer                            447
                                  ... 
Full Cover,Full                      1
combinational,sensitive              1
Lightweight Formula,Full             1
Face                                 1
fine, damaged                        1
Name: count, Length: 134, dtype: int64

##### 'type' Column 
because of 2681 NaN is mixed-type columns and I will  replaced it with Mode

In [11]:
df_cosmetic['color'].value_counts(dropna = False)

color
No color            4197
NaN                 1989
Black                440
Multicolor           192
No colour            161
                    ... 
Snowball               1
5. Lustrous Nude       1
Certainly Red          1
Power Move             1
Selfie Queen           1
Name: count, Length: 3358, dtype: int64

##### 'color' Column
have many false entered data such as NaN, #Error, 9#, 4#, long text,... so I will removed this column.

In [12]:
df_cosmetic['size'].value_counts(dropna = False)

size
NaN            3166
30              835
100             435
200             349
50              317
               ... 
1.4 s             1
0.19 Pounds       1
13 s              1
0.14 Kilos        1
47                1
Name: count, Length: 701, dtype: int64

##### 'Size' Column
The same unit for the size is not mentioned and could not be also estimated, therefore should this column be removed

In [13]:
df_cosmetic['rating'].value_counts(dropna = False)

rating
NaN     2067
4.3      757
4.2      714
4.5      692
4.4      689
        ... 
1.7        1
85         1
53         1
54         1
3.95       1
Name: count, Length: 254, dtype: int64

##### 'rating' Column
has NaN that will be replaced with mean, and Data more than 5, als datas more than five will be corrected and replace with 5, type of the column will be corrected from object to floating point and these type of data:'2.8 out of 5 stars' and '3.5 out of 5 stars' will be replaced with 2.8 and 3.5 

In [14]:
df_cosmetic['noofratings'].value_counts(dropna = False)

noofratings
0             686
No reviews    513
NaN           459
1             322
14            318
             ... 
12,653          1
6,191           1
3,985           1
3,209           1
587             1
Name: count, Length: 2097, dtype: int64

##### 'noofratings' Column
Column name will be renamed to Count of ratings
the word of 'Ratings' in this column will be removed and the number will be remained. 'New to Amazon', 'No reviews', 'write A review' and 'NaN' will be replaced  with'0'
type of the column will be Integer.

In [15]:
# finding missing values
df_cosmetic.isnull().sum()

website           0
country           0
category          0
subcategory       0
price           317
brand             0
form              0
type           2681
color          1989
size           3166
rating         2067
noofratings     459
dtype: int64

In [16]:
# find duplicates
df_dups = df_cosmetic[df_cosmetic.duplicated()]
df_dups

website country category subcategory   price                   brand  \
59        ulta     USA     body    bodywash  1634.0              Coco & Eve   
68        ulta     USA     body    bodywash   653.6                 Da Bomb   
72        ulta     USA     body    bodywash   653.6                 Da Bomb   
73        ulta     USA     body    bodywash   653.6                 Da Bomb   
78        ulta     USA     body    bodywash   653.6                 Da Bomb   
...        ...     ...      ...         ...     ...                     ...   
11835     ulta     USA     face  foundation  1143.8  ULTA Beauty Collection   
12132  Sephora   India     face   concealer  1450.0                Smashbox   
12532  Sephora   India     face   concealer  2000.0      Sephora Collection   
12533  Sephora   India     face   concealer  2000.0      Sephora Collection   
12538  Sephora   India     face       blush  2400.0                    PIXI   

         form                            type        color  size rating  \
59      cream                             all     No color     0    NaN   
68      cream                             all     No color   210    NaN   
72      cream                             all     No color   210    NaN   
73      cream                             all     No color   210      5   
78      cream                             all     No color   210    NaN   
...       ...                             ...          ...   ...    ...   
11835  powder                 dry,oily,normal  Tan Neutral  10.5    4.2   
12132   cream  Combination, Dry, Oily, Normal    No colour    10   3.14   
12532   serum  Combination, Dry, Normal, Oily    No colour    30    3.1   
12533   serum  Combination, Dry, Normal, Oily    No colour    30    3.1   
12538   serum  Combination, Dry, Normal, Oily    No colour    30   3.11   

      noofratings  
59            NaN  
68            NaN  
72            NaN  
73              2  
78            NaN  
...           ...  
11835         322  
12132          14  
12532           9  
12533           9  
12538          11  

[602 rows x 12 columns]

In [17]:
df_dups.shape

(602, 12)

### I will remove duplicates first

In [18]:
#remove duplicates in-place
df_cosmetic = df_cosmetic.drop_duplicates()

In [19]:
# check duplicates again
df_dups_1 = df_cosmetic[df_cosmetic.duplicated()]
df_dups_1

Empty DataFrame
Columns: [website, country, category, subcategory, price, brand, form, type, color, size, rating, noofratings]
Index: []

In [20]:
# Columns'Product_name', 'title_hef' ,'ingredients', 'color' and 'size'are removed
vars_list = ['website', 'country', 'category', 'subcategory', 'price', 'brand', 'form', 'type', 'rating', 'noofratings']
df_cosmetic = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Ecommerce_cosmetic_dataset.csv'),usecols=vars_list, encoding='latin1', index_col=False)

### 'rating' Column correction

In [21]:
df_cosmetic['rating'] = df_cosmetic['rating'].str.replace(',', '.')

In [22]:
df_cosmetic['rating'] = df_cosmetic['rating'].str.replace(' out of 5 stars', '')

In [23]:
# Changing data type of rating
df_cosmetic['rating'] = df_cosmetic['rating'].astype('float')
df_cosmetic['rating'] 

0        3.90
1        4.40
2        4.40
3        4.20
4        4.30
         ... 
12610    3.15
12611    3.13
12612    4.03
12613    3.15
12614    3.16
Name: rating, Length: 12615, dtype: float64

In [24]:
#replacing NaN values with the mean
mean_value_rating = df_cosmetic['rating'].mean()
df_cosmetic['rating'].fillna(mean_value_rating, inplace=True)

In [25]:
# Replace values greater than 5 with 5:
df_cosmetic.loc[df_cosmetic['rating'] > 5, 'rating'] = 5

In [26]:
df_cosmetic['rating'].value_counts(dropna = False)

rating
4.866494    2067
4.300000     781
4.200000     739
4.500000     713
4.400000     703
            ... 
4.690000       1
4.780000       1
4.640000       1
4.480000       1
3.950000       1
Name: count, Length: 181, dtype: int64

### 'noofratings' corrections

In [27]:
df_cosmetic['noofratings'] = df_cosmetic['noofratings'].str.replace(',', '')

In [28]:
# Clean the column
df_cosmetic['noofratings'] = df_cosmetic['noofratings'].str.replace(',', '')
df_cosmetic['noofratings'] = df_cosmetic['noofratings'].str.replace('Ratings', '')
df_cosmetic['noofratings'] = df_cosmetic['noofratings'].str.replace('ratings', '')
df_cosmetic['noofratings'].replace({'New to Amazon': '0', 'No reviews': '0', 'Write A Review': '0'}, inplace=True)
df_cosmetic['noofratings'] = df_cosmetic['noofratings'].fillna('0')  # Ensure NaN is replaced with string '0'


# Convert to integer type
df_cosmetic['noofratings']= df_cosmetic['noofratings'].astype(int)

In [29]:
df_cosmetic['noofratings'].value_counts(dropna = False)

noofratings
0       1876
14       323
1        322
11       321
7        306
        ... 
1220       1
1169       1
3290       1
1699       1
587        1
Name: count, Length: 1906, dtype: int64

### 'type' column correction

In [30]:
# Calculate the mode of the 'type' column
mode_value_type = df_cosmetic['type'].mode()[0]

# Replace NaN values with the mode
df_cosmetic['type'] = df_cosmetic['type'].fillna(mode_value_type)

In [31]:
df_cosmetic['type'].value_counts(dropna = False)

type
all                               7839
matte                              572
combination, dry, normal, oily     528
shimmer                            447
Combination, Dry, Normal, Oily     381
                                  ... 
Full Cover,Full                      1
combinational,sensitive              1
Lightweight Formula,Full             1
Face                                 1
fine, damaged                        1
Name: count, Length: 133, dtype: int64

In [32]:
# finding missing values after data cleaning
df_cosmetic.isnull().sum()

website          0
country          0
category         0
subcategory      0
price          317
brand            0
form             0
type             0
rating           0
noofratings      0
dtype: int64

### I will replace missing value of 'price' with mean

In [33]:
# Calculate the mean of the 'price' column
mean_value_price = df_cosmetic['price'].mean()

# Replace NaN values with the mean
df_cosmetic['price'] = df_cosmetic['price'].fillna(mean_value_price)

In [34]:
df_cosmetic.shape

(12615, 10)

## 4. Renaming Columns

In [35]:
df_cosmetic.columns

Index(['website', 'country', 'category', 'subcategory', 'price', 'brand',
       'form', 'type', 'rating', 'noofratings'],
      dtype='object')

In [36]:
# Rename the 'noofratings'column
df_cosmetic.rename(columns = {'noofratings' : 'count of ratings'}, inplace = True)

In [37]:
df_cosmetic.columns

Index(['website', 'country', 'category', 'subcategory', 'price', 'brand',
       'form', 'type', 'rating', 'count of ratings'],
      dtype='object')

## 5. Data Understanding

In [38]:
# basic descriptive statistical analysis
df_cosmetic.describe()

price        rating  count of ratings
count  12615.000000  12615.000000      12615.000000
mean    2281.180935      4.257113        994.659770
std     3118.747543      0.556765       6570.809313
min        1.700000      1.000000          0.000000
25%      499.000000      4.060000          6.000000
50%     1390.000000      4.300000         26.000000
75%     2799.500000      4.700000        266.500000
max    94099.000000      5.000000     220040.000000

#### Analysis makes sense

In [42]:
df_cosmetic.shape

(12615, 10)

## Exporting dataframe

In [40]:
# export dataframe:
df_cosmetic.to_csv(os.path.join(path, '02 Data','Prepared Data', 'cosmetic_clean.csv'))